replica/table.cc

```diff
@@ -550,9 +550,27 @@ void table::enable_off_strategy_trigger() {
 
 storage_group_manager::~storage_group_manager() = default;
 
+future<> storage_group_manager::parallel_foreach_storage_group(std::function<future<>(storage_group&)> f) {
+    auto storage_groups = _storage_groups;
+    co_await coroutine::parallel_for_each(storage_groups, [&] (std::pair<size_t, storage_group_ptr> p) -> future<> {
+        auto& sg = p.second;
+        // if table wide op (e.g. split) finds group that was closed by migration cleanup, it will be skipped.
+        if (sg->async_gate().is_closed()) {
+            co_return;
+        }
+        auto holder = sg->async_gate().hold();
+        co_await f(*sg);
+    });
+    co_return;
+}
+
 future<> storage_group_manager::for_each_storage_group_gently(std::function<future<>(size_t, storage_group&)> f) {
-    rwlock::holder shared_lock = co_await get_rwlock().hold_read_lock();
-    for (auto& [id, sg]: _storage_groups) {
+    auto storage_groups = _storage_groups;
```

Comment:
here's the copy is justified since we may yield in the loop body

Reply:
If we copy storage_groups map, preempt and the map will be modified, `storage_group_manager::_storage_groups` and `storage_group_manager::_compaction_groups` may become inconsistent: 
a storage group is removed, while compaction group is still there (`storage_group_manager::_compaction_groups` is an intrusive list with auto-unlink).

Reply:
We need to remove this intrusive list. Probably a good opportunity

Reply:
done